django/db/models/sql/compiler.py

```diff
@@ -304,10 +304,10 @@ def get_order_by(self):
                 if not self.query.standard_ordering:
                     field = field.copy()
                     field.reverse_ordering()
-                order_by.append((field, False))
+                order_by.append((field, False, field))
                 continue
             if field == '?':  # random
-                order_by.append((OrderBy(Random()), False))
+                order_by.append((OrderBy(Random()), False, field))
```

Comment:
Are these cases covered by tests?

In all other cases `col` is actually `str: alias` but in these cases it's `Expression: expr` or a `str` that doesn't point to a column. Given this value will end up in `Query.values_select` which is `tuple[str]` through the `add_select_col` call I think there might be something wrong here?

I guess it could be reproduced through something like `order_by(F('annotated_field') * 2, ...)`?

Reply:
You're right, the current approach is pretty wrong 👍 

The case of `order_by(F('annotated_field') * 2, ...)` is still buggy, and the same goes for more operations using two annotated fields for instance: `order_by(F('annotated_field') * F('some_other_field'), ...)`.
I'll rework the PR and try to come up with something 👌 

But the fact that `expr.resolve_expression(self.query, allow_joins=True, reuse=None)` with expr = `OrderBy(F(annotated_field) * F(some_other_field), descending=False)` will still return a `OrderBy(Value(2) * Value(2), descending=False)` even in the case of a "combinator" query feels like just making an edge case work through a workaround, rather than fixing the general approach.

So on one hand, it feels like `resolve_expression` should be aware of the query type, and resolve that `F(annotated_field)` is not just `Value(2)`, but the annotated alias because we are in the context of a combinator query (so it cannot just go down to the hard-coded value of the first QuerySet).
On the other hand, this also feels like if I start editing the `resolve_expression` method, I will have to update potentially all those methods, which are rather numerous 🤔 

I'll dig into it 🕳️ 💪